<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Open-the-table-and-examine-the-general-information-about-the-data" data-toc-modified-id="Open-the-table-and-examine-the-general-information-about-the-data-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Open the table and examine the general information about the data</a></span></li><li><span><a href="#Data-preprocessing" data-toc-modified-id="Data-preprocessing-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data preprocessing</a></span><ul class="toc-item"><li><span><a href="#Remove-gaps" data-toc-modified-id="Remove-gaps-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Remove gaps</a></span></li><li><span><a href="#Handling-anomalous-values" data-toc-modified-id="Handling-anomalous-values-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Handling anomalous values</a></span></li><li><span><a href="#Removing-gaps-(continued)" data-toc-modified-id="Removing-gaps-(continued)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Removing gaps (continued)</a></span></li><li><span><a href="#Changing-data-types" data-toc-modified-id="Changing-data-types-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Changing data types</a></span></li><li><span><a href="#Handling-duplicates" data-toc-modified-id="Handling-duplicates-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Handling duplicates</a></span></li><li><span><a href="#Data-categorization" data-toc-modified-id="Data-categorization-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Data categorization</a></span></li><li><span><a href="#Explore-data-and-answer-questions" data-toc-modified-id="Explore-data-and-answer-questions-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Explore data and answer questions</a></span><ul class="toc-item"><li><span><a href="#Is-there-a-relationship-between-the-number-of-children-and-loan-repayment-on-time?" data-toc-modified-id="Is-there-a-relationship-between-the-number-of-children-and-loan-repayment-on-time?-2.7.1"><span class="toc-item-num">2.7.1&nbsp;&nbsp;</span>Is there a relationship between the number of children and loan repayment on time?</a></span></li><li><span><a href="#Is-there-a-relationship-between-marital-status-and-loan-repayment-on-time?" data-toc-modified-id="Is-there-a-relationship-between-marital-status-and-loan-repayment-on-time?-2.7.2"><span class="toc-item-num">2.7.2&nbsp;&nbsp;</span>Is there a relationship between marital status and loan repayment on time?</a></span></li><li><span><a href="#Is-there-a-relationship-between-income-level-and-loan-repayment-on-time?" data-toc-modified-id="Is-there-a-relationship-between-income-level-and-loan-repayment-on-time?-2.7.3"><span class="toc-item-num">2.7.3&nbsp;&nbsp;</span>Is there a relationship between income level and loan repayment on time?</a></span></li><li><span><a href="#How-do-different-purposes-of-a-loan-affect-its-repayment-on-time?" data-toc-modified-id="How-do-different-purposes-of-a-loan-affect-its-repayment-on-time?-2.7.4"><span class="toc-item-num">2.7.4&nbsp;&nbsp;</span>How do different purposes of a loan affect its repayment on time?</a></span></li><li><span><a href="#Give-possible-reasons-for-gaps-in-the-original-data." data-toc-modified-id="Give-possible-reasons-for-gaps-in-the-original-data.-2.7.5"><span class="toc-item-num">2.7.5&nbsp;&nbsp;</span>Give possible reasons for gaps in the original data.</a></span></li><li><span><a href="#Explain-why-filling-in-the-gaps-with-the-median-is-the-best-solution-for-scale-variables." data-toc-modified-id="Explain-why-filling-in-the-gaps-with-the-median-is-the-best-solution-for-scale-variables.-2.7.6"><span class="toc-item-num">2.7.6&nbsp;&nbsp;</span>Explain why filling in the gaps with the median is the best solution for scale variables.</a></span></li></ul></li><li><span><a href="#General-conclusion" data-toc-modified-id="General-conclusion-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>General conclusion</a></span></li></ul></li></ul></div>

# Research of reliability of borrowers

## Open the table and examine the general information about the data

**Task 1. Import the `pandas` library. Read the data from the csv file into a dataframe and store it in the `data` variable. The path to the file:**

`/datasets/data.csv`

In [1]:
import pandas as pd
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Task 2. Display the first 20 lines of dataframe `data` on the screen.**

In [2]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

**Task 3. Display basic dataframe information using the `info()` method.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Data preprocessing

### Remove gaps

**Task 4. Print the number of missing values ​​for each column. Use a combination of the two methods.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Task 5. There are missing values ​​in two columns. One of them is `days_employed`. The gaps in this column will be dealt with in the next step. Another column with missing values, `total_income`, stores income data. The amount of income is most affected by the type of employment, so fill in the gaps in this column with the median value for each type from the `income_type` column. For example, for a person with an employment type of `employee`, the gap in the `total_income` column must be filled by the median income among all records with the same type.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Handling anomalous values

**Task 6. Artifacts (anomalies) may occur in the data - values ​​that do not reflect reality and appeared due to some kind of error. such an artifact would be the negative number of days of work experience in the `days_employed` column. For real data, this is normal. Process the values ​​in this column: replace all negative values ​​with positive ones using the `abs()` method.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Task 7. For each type of employment, print the median `days_employed` in days.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

Two types (unemployed and pensioners) will get abnormally large values. Correcting such values ​​is difficult, so leave them as they are. Moreover, you will not need this column for research.

**Task 8. List the unique values ​​of the `children` column.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Task 9. There are two abnormal values ​​in the `children` column. Remove lines containing such anomalous values ​​from the `data` dataframe.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Task 10. List the unique values ​​of the `children` column again to make sure the artifacts have been removed.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Removing gaps (continued)

**Task 11. Fill in the gaps in the `days_employed` column with the median values ​​for each `income_type` employment type.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Task 12. Make sure all gaps are filled. Test yourself and print the number of missing values for each column again using two methods.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Changing data types

**Task 13. Replace the real data type in the `total_income` column with an integer using the `astype()` method.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Handling duplicates

**Task 14. Handle implicit duplicates in the `education` column. This column has the same values, but written differently: using uppercase and lowercase letters. Convert them to lower case. Check other columns.**

In [14]:
data['education'] = data['education'].str.lower()

**Task 15. Display the number of duplicate rows in the data. If such lines are present, remove them.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Data categorization

**Task 16. Based on the ranges below, create a `total_income_category` column with categories in the `data` dataframe:**

- 0–30000 - `'E'`;
- 30001–50000 - `'D'`;
- 50001–200000 - `'C'`;
- 200001–1000000 - `'B'`;
- 1000001 and above - `'A'`.


**For example, a borrower with an income of 25,000 should be assigned an `'E'` category, and a customer with an income of 235,000 should be assigned a `'B'` category. Use your own function named `categorize_income()` and method `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Task 17. Display a list of unique loan purposes from the `purpose` column.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Task 18. Create a function that, based on the data from the `purpose` column, will form a new `purpose_category` column, which will include the following categories:**

- `'car operations'`,
- `'real estate transactions'`,
- `'conducting a wedding'`,
- `'getting an education'`.

**For example, if the `purpose` column contains the substring `'car purchase'`, then the `purpose_category` column should contain the string `'car operations'`.**

**Use your own function named `categorize_purpose()` and method `apply()`. Examine the data in the `purpose` column and determine which substrings will help you correctly identify the category.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'car operations'
        elif 'жил' in row or 'недвиж' in row:
            return 'real estate transactions'
        elif 'свад' in row:
            return 'conducting a wedding'
        elif 'образов' in row:
            return 'getting an education'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Explore data and answer questions

#### Is there a relationship between the number of children and loan repayment on time?

In [22]:
def data_pivot(category):
    return data.pivot_table(['debt'], index=[category], aggfunc = ['count', 'sum', 'mean'])\
.rename(columns={'count':'Total borrowers', 'sum':'Total debtors', 'mean':'Share of debtors'}).droplevel(1,axis=1)
data_pivot(data['children'])

Total borrowers  Total debtors  Share of debtors
children                                                  
0                   14091           1063          0.075438
1                    4808            444          0.092346
2                    2052            194          0.094542
3                     330             27          0.081818
4                      41              4          0.097561
5                       9              0          0.000000

**Conclusion:** People without children are less likely to have accounts receivable, only 7.54%. People with three children are slightly more likely, 8.18%. It is difficult to say about people with four and five children, since out of the entire sample only 9 people have 5 children, 41 people have 4 children, and in our sample these are very small values ​​for a full assessment.
Therefore, the most reliable payers are people without children.

#### Is there a relationship between marital status and loan repayment on time?

In [23]:
data_pivot(data['family_status'])

Total borrowers  Total debtors  Share of debtors
family_status                                                          
Не женат / не замужем             2796            273          0.097639
в разводе                         1189             84          0.070648
вдовец / вдова                     951             63          0.066246
гражданский брак                  4134            385          0.093130
женат / замужем                  12261            927          0.075606

**Conclusion:** The most reliable borrowers are widows / widowers and people who are divorced, the most unreliable are single / unmarried

#### Is there a relationship between income level and loan repayment on time?

In [24]:
data_pivot(data['total_income_category'])

Total borrowers  Total debtors  Share of debtors
total_income_category                                                  
A                                   25              2          0.080000
B                                 5014            354          0.070602
C                                15921           1353          0.084982
D                                  349             21          0.060172
E                                   22              2          0.090909

**Conclusion:** The most reliable borrowers are people from the B category, i.e. with an income of 200,001 to 1,000,000, a little less reliable - from the C category, i.e. with an income of 50,001 to 200,000, it is difficult to judge for other categories due to lack of data

#### How do different purposes of a loan affect its repayment on time?

In [25]:
data_pivot(data['purpose_category'])

Total borrowers  Total debtors  Share of debtors
purpose_category                                                          
car operations                       4279            400          0.093480
conducting a wedding                 2313            183          0.079118
getting an education                 3988            369          0.092528
real estate transactions            10751            780          0.072551

**Conclusion:** Least of all debts on loans for the purpose of real estate transactions, most of all - for loans for the purpose of car transactions

#### Give possible reasons for gaps in the original data.

*Answer:* A human factor is possible, for example, the client did not provide supporting documents, perhaps they were not required (for example, some data may not be required if the client has a good credit history in this bank or the loan amount is small), or there could be an error of the employee who filled in the data. A technical factor is also possible - some kind of failure occurred, an error occurred during the transfer, merging, unloading or loading of data.

#### Explain why filling in the gaps with the median is the best solution for scale variables.

*Answer:* Because the median value allows you to smooth out the influence of values ​​that stand out from the total sample.

### General conclusion

Depending on the probability of the client repaying the loan on time, you can adjust the interest rate of the loan product, as well as issue a refusal to a potential borrower in case of unreliability.
The analysis of data with statistics of the solvency of customers was carried out, gaps in the data were removed, anomalous values ​​were processed, duplicates were processed. Also, the categorization of the data was made and the relationship between the number of children, marital status, the purpose of obtaining a loan, income level and repayment of the loan on time was investigated.
We found that the probability of repaying a loan on time changes as follows:
- the probability of debt is low (7.54%) if the borrower has no children, and higher if the borrower has 1 or 2 children (9.23% and 9.45% respectively)
- the most reliable borrowers by marital status are widows / widowers (probability of debt 6.62%) and divorced people (probability of debt 7.06%), the most unreliable - single / unmarried (probability of debt 9.76%)
- the most reliable borrowers in terms of income - people with income from 200,001 to 1,000,000 (probability of debt 7.06%), slightly less reliable - with income from 50,001 to 200,000 (probability of debt 8.49%), it is difficult to judge for other categories due to lack of data
- the smallest number of debts on loans for the purpose of real estate transactions (7.25%), the most - for loans for the purpose of car transactions (9.34%)

Thus, the most reliable borrower is a person without children, widower / widow or divorced, with an income of 200,001 to 1,000,000, who takes out a loan for a real estate transaction.